### PR TITLE
[system-health] Add support for monitoring system health

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
**- Why I did it**

System health feature requires a service to collect system status and manage the system status LED.

**- How I did it**

1. Add a health-check python library at src/system-health/health_check. This library is called by healthd and CLI to get system status information
2. Add a healthd daemon to run as a service on host side. The daemon collects system status information at a configurable interval and puts results to state DB.
3. Add make files and service file 

**- How to verify it**

Manual test on SN2700. Regression test on SN2700, SN4600C, SN4700, SN3800, SN2100.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: slave.mk
Comment (sujinmkang) [inline slave.mk:790]: please change the tab to spaces

**Comment by sujinmkang** [comment]
Can we move this to pmon container? The hardware checker part is monitoring the platform status. I think pmon could be the right place to monitor. @jleveque  what do you think?

**Comment by jleveque** [comment]
@sujinmkang: While I agree that the hardware_checker.py makes sense to reside in the PMon container, the rest of the modules do not, as they check the status of services, etc., they belong in the host. Therefore, it would be difficult (and a bit messy) to separate this in to multiple components, some which reside in the host, and one of which resides in the PMon container. For now, I say we keep this in one location, and consider how we could possibly separate it in the future.

File: slave.mk
Comment (jleveque) [reply slave.mk:790]: Please check all files for similar whitespace issues, as well.

Comment (Junchao-Mellanox) [reply slave.mk]: Fixed

File: src/system-health/health_checker/manager.py
Comment (sujinmkang) [inline src/system-health/health_checker/manager.py:24]: If the platform config file does not exist, what is the expected action next?

File: src/system-health/health_checker/system_health_monitoring_config.json
Comment (sujinmkang) [inline src/system-health/health_checker/system_health_monitoring_config.json:11]: Is this a sample configuration file or the default configuration file?

Comment (Junchao-Mellanox) [reply src/system-health/health_checker/system_health_monitoring_config.json:11]: It is a sample configuration file.

File: src/system-health/health_checker/manager.py
Comment (Junchao-Mellanox) [reply src/system-health/health_checker/manager.py:24]: We have default value for all configuration entries. If config file does not exist, default value will be used.

**Comment by keboliu** [comment]
@sujinmkang @jleveque  would you please check the response from Junchao and see if still have further comments?

**Comment by liat-grozovik** [comment]
@jleveque should we wait for your approval or we can move on with merging it?

**Comment by shlomibitton** [comment]
@liat-grozovik as we discussed, the CLI reference is on sonic-utilities:
https://github.com/Azure/sonic-utilities/pull/971

**Comment by jleveque** [comment]
Please fix conflicts.

File: rules/system-health.mk
Comment (jleveque) [inline rules/system-health.mk:1]: Can you please also add a `system-health.dep` file for our new package caching system? You should be able to easily copy an existing file for another whl, like https://github.com/Azure/sonic-buildimage/blob/master/rules/sonic_bgpcfgd.dep and make minor changes.

File: src/system-health/health_checker/config.py
Comment (jleveque) [inline src/system-health/health_checker/config.py:4]: sonic_device_util is being deprecated in favor of the new sonic-py_common package. Please update accordingly. See https://github.com/Azure/sonic-buildimage/pull/5043 for reference.

File: src/system-health/scripts/healthd
Comment (jleveque) [inline src/system-health/scripts/healthd:11]: sonic_daemon_base is being deprecated in favor of the new sonic-py_common package. Please update accordingly. See https://github.com/Azure/sonic-buildimage/pull/5043 and https://github.com/Azure/sonic-platform-daemons/pull/74 for reference.

File: src/system-health/tests/test_system_health.py
Comment (jleveque) [inline src/system-health/tests/test_system_health.py:10]: Remove commented code

File: src/system-health/tests/test_system_health.py
Comment (jleveque) [inline src/system-health/tests/test_system_health.py:21]: Remove commented code

File: rules/system-health.mk
Comment (Junchao-Mellanox) [reply rules/system-health.mk:1]: Sure.

File: src/system-health/health_checker/config.py
Comment (Junchao-Mellanox) [reply src/system-health/health_checker/config.py:4]: Sure

File: src/system-health/scripts/healthd
Comment (Junchao-Mellanox) [reply src/system-health/scripts/healthd:11]: Sure

File: src/system-health/tests/test_system_health.py
Comment (Junchao-Mellanox) [reply src/system-health/tests/test_system_health.py:21]: Sure

**Comment by Junchao-Mellanox** [comment]
> Please fix LGTM alert. If you want to catch all exceptions, use `except Exception:`

Fixed

**Comment by keboliu** [comment]
retest mellanox please

**Comment by Junchao-Mellanox** [comment]
retest vsimage please

**Comment by Junchao-Mellanox** [comment]
retest mellanox please

**Comment by Junchao-Mellanox** [comment]
retest vs please

**Comment by keboliu** [comment]
retest vs please

**Comment by keboliu** [comment]
retest vsimage please

File: files/image_config/system-health/system-health.service
Comment (jleveque) [inline files/image_config/system-health/system-health.service:2]: s/Monitor system health/SONiC system health monitro/

File: src/system-health/setup.py
Comment (jleveque) [inline src/system-health/setup.py:23]: Can you add an `install_requires` section here and list all the wheels needed at runtime? E.g.,
```
      install_requires=[
          'sonic-py-common',
          'swsssdk>=2.0.1'
      ],
```

File: src/system-health/tests/test_system_health.py
Comment (jleveque) [inline src/system-health/tests/test_system_health.py:4]: Add a newline below this

File: src/system-health/tests/test_system_health.py
Comment (jleveque) [inline src/system-health/tests/test_system_health.py:2]: Add a newline below this

File: src/system-health/setup.py
Comment (Junchao-Mellanox) [reply src/system-health/setup.py:23]: Sure.

**Comment by Junchao-Mellanox** [comment]
retest mellanox please

**Comment by liat-grozovik** [comment]
retest baseimage please

**Comment by Junchao-Mellanox** [comment]
retest vs please

File: src/system-health/health_checker/config.py
Comment (jleveque) [inline src/system-health/health_checker/config.py:45]: Please rename `user_defined_checkers`, and update all references throughout all files.

File: src/system-health/health_checker/user_define_checker.py
Comment (jleveque) [inline src/system-health/health_checker/user_define_checker.py:8]: s/user define/user-defined/

File: device/mellanox/x86_64-mlnx_msn2010-r0/system_health_monitoring_config.json
Comment (jleveque) [inline device/mellanox/x86_64-mlnx_msn2010-r0/system_health_monitoring_config.json:4]: This should also be changed to `user_defined_checkers`

File: device/mellanox/x86_64-mlnx_msn2700-r0/system_health_monitoring_config.json
Comment (jleveque) [inline device/mellanox/x86_64-mlnx_msn2700-r0/system_health_monitoring_config.json:4]: This should also be changed to `user_defined_checkers`

File: device/mellanox/x86_64-mlnx_msn2700_simx-r0/system_health_monitoring_config.json
Comment (jleveque) [inline device/mellanox/x86_64-mlnx_msn2700_simx-r0/system_health_monitoring_config.json:4]: This should also be changed to `user_defined_checkers`

File: src/system-health/health_checker/system_health_monitoring_config.json
Comment (jleveque) [inline src/system-health/health_checker/system_health_monitoring_config.json:4]: This should also be changed to `user_defined_checkers`

File: src/system-health/health_checker/user_define_checker.py
Comment (jleveque) [inline src/system-health/health_checker/user_define_checker.py:1]: This filename should also be updated to `user_defined_checker.py`

File: src/system-health/health_checker/manager.py
Comment (jleveque) [inline src/system-health/health_checker/manager.py:48]: This also needs to be updated to `user_defined_checker`

File: src/system-health/tests/test_system_health.py
Comment (jleveque) [inline src/system-health/tests/test_system_health.py:33]: This also needs to be updated to `user_defined_checker`

**Comment by Junchao-Mellanox** [comment]
retest baseimage please

File: src/system-health/health_checker/config.py
Comment (jleveque) [inline src/system-health/health_checker/config.py:3]: Group imports per [PEP8 standards](https://www.python.org/dev/peps/pep-0008/#imports), and alphabetize within groups:

```suggestion
import json
import os

from sonic_py_common import device_info
```

File: src/system-health/health_checker/hardware_checker.py
Comment (jleveque) [inline src/system-health/health_checker/hardware_checker.py:3]: Add blank line to separate third-party from local imports:

```suggestion
from natsort import natsorted
from swsssdk import SonicV2Connector

from .health_checker import HealthChecker
```

File: src/system-health/scripts/healthd
Comment (jleveque) [inline src/system-health/scripts/healthd:13]: Reorganize imports:

```suggestion
import signal
import threading

from sonic_py_common.daemon_base import DaemonBase
from sonic_py_common.logger import Logger
from swsssdk import SonicV2Connector

from health_checker.manager import HealthCheckerManager
```

File: src/system-health/tests/test_system_health.py
Comment (jleveque) [inline src/system-health/tests/test_system_health.py:24]: Do the last two imports need to come after the path setup? If not, can you move to the top and group/organize imports per PEP8 standards?

File: src/system-health/tests/test_system_health.py
Comment (jleveque) [inline src/system-health/tests/test_system_health.py:59]: These are imported in a couple test functions. Why not import at the top of the file?

File: src/system-health/health_checker/config.py
Comment (Junchao-Mellanox) [reply src/system-health/health_checker/config.py:3]: Hi, thanks for sharing the standard. I will fix them soon.

File: src/system-health/tests/test_system_health.py
Comment (jleveque) [inline src/system-health/tests/test_system_health.py:18]: Add newline below this import

**Comment by jleveque** [comment]
LGTM. @sujinmkang to review again, please.

**Review by keboliu — DISMISSED** [review]

File: src/system-health/health_checker/manager.py
Comment (Junchao-Mellanox) [reply src/system-health/health_checker/manager.py:41]: 1. _can we make sure if the default configuration works for all other platforms?_
It depends on how we define "works". Like you said, there might be platforms doesn't support setting system led or lack of some information in DB, per the default configuration, it would either trigger a warning syslog or an error message in the output of "show system-health summary" command. 
2. _The default boot up time is 300s which is 5mins. Until uptime passes 5 mins, does it stay in booting state even for the case that all the service and daemon are already running?_
Yes. Now system monitor service is depending on monit service to determine the status of critical services/process/file systems. monit service has a configuration file which set the start delay to 300s for itself. System monitor service will always try reading that config file to determine the wait time. If monit configuration file doesn't exist, it will use default 300s.
3. _why this system health monitoring service needs to set the system health led?_
This is according to the design. The system led is a physical component that using for indicating the status of the switch, so if there is anything wrong in the system, the led status should be red to inform user that something bad is happending.
4. _It will be safe to disable this health monitoring service unless the platform specific configuration file exists_
Yes, it is safer to do that. For now, only mellanox platform has such configuration file. @jleveque what's your opinion about this?

File: src/system-health/health_checker/manager.py
Comment (Junchao-Mellanox) [reply src/system-health/health_checker/manager.py:90]: System monitor service is depending on monit service, so it needs to wait monit service to start working. So the boot time out here is actually getting from monit service configuration file located at /etc/monit/monitrc. I also wanted to monitor some expected services and/or containers, but I found that there is no such definition of "boot up done" for all platforms now. So we are using a boot time out to simplify the implementation for now.

File: src/system-health/health_checker/manager.py
Comment (Junchao-Mellanox) [reply src/system-health/health_checker/manager.py:41]: Added a check to system health configuration file, if the file does not exist, healthd just exit.

**Comment by Junchao-Mellanox** [comment]
@sujinmkang could you please review again? Thanks.

**Comment by Junchao-Mellanox** [comment]
retest mellanox please

**Comment by Junchao-Mellanox** [comment]
retest broadcom please

**Comment by Junchao-Mellanox** [comment]
retest mellanox please

**Comment by Junchao-Mellanox** [comment]
retest broadcom please

**Comment by sujinmkang** [comment]
retest mellanox please

**Comment by sujinmkang** [comment]
retest broadcom please

**Review by keboliu — DISMISSED** [review]

**Comment by Junchao-Mellanox** [comment]
retest vsimage please

**Comment by Junchao-Mellanox** [comment]
retest broadcom please

File: src/system-health/health_checker/system_health_monitoring_config.json
Comment (jleveque) [inline src/system-health/health_checker/system_health_monitoring_config.json:11]: Please use 4 spaces to indent

**Review by jleveque — CHANGES_REQUESTED** [review]

**Review by sujinmkang — DISMISSED** [review]

File: src/system-health/pytest.ini
Comment (jleveque) [inline src/system-health/pytest.ini:3]: I think we can do away with this warning.

Also, I suggest adding options to print out test coverage reports. Please use the following as an example: https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-py-common/pytest.ini

**Review by jleveque — CHANGES_REQUESTED** [review]

File: src/system-health/pytest.ini
Comment (Junchao-Mellanox) [reply src/system-health/pytest.ini]: Fixed

File: src/system-health/health_checker/system_health_monitoring_config.json
Comment (Junchao-Mellanox) [reply src/system-health/health_checker/system_health_monitoring_config.json]: Fixed

File: src/system-health/pytest.ini
Comment (jleveque) [inline src/system-health/pytest.ini:2]: Looks like this was copy/pasted but not modified, as it still references `sonic_py_common`

**Review by jleveque — CHANGES_REQUESTED** [review]

File: src/system-health/pytest.ini
Comment (Junchao-Mellanox) [reply src/system-health/pytest.ini]: Fixed

File: src/system-health/pytest.ini
Comment (jleveque) [inline src/system-health/pytest.ini:2]: Does this also report coverage for `healthd`? Maybe you need to add an extra `--cov=scripts`?

Comment (Junchao-Mellanox) [reply src/system-health/pytest.ini:2]: The current unit test doesn't test healthd, because healthd has a dead loop and most of the logic is in health_checker.

Comment (jleveque) [reply src/system-health/pytest.ini:2]: OK. If healthd is ever modified such that tests can be added, we'll add it to the coverage at that time.

File: src/system-health/health_checker/manager.py
Comment (jleveque) [reply src/system-health/health_checker/manager.py:41]: I think it is fine for now to log a message and exit if the config file doesn't exist, as you're now doing.

File: src/system-health/health_checker/manager.py
Comment (jleveque) [reply src/system-health/health_checker/manager.py:90]: Even if we eventually have a definition of "boot up done" for all platforms, we will still need some sort of max timeout here, in case the device never reaches "boot up done" for some reason.

To Sujin's point, it would be optimal if we could set this timeout based on the type of boot being performed. One approach I can think of is to have the different timeout values in the config file and choose on based on the type of boot being performed based on the previous reboot cause.

Comment (Junchao-Mellanox) [reply src/system-health/health_checker/manager.py:90]: As system health service is designed to depend on monit service to check the service/process/file system status, so system health always waits for monit service. I did some tests with monit service and found that monit service always wait 5 minutes after the reboot no matter what the reboot type is, so system health service have to wait 5 minutes here for all reboot type.

Comment (jleveque) [reply src/system-health/health_checker/manager.py:90]: Understood. For us to use a variable timeout here, we would first need to implement a variable delay for Monit.

File: src/system-health/scripts/healthd
Comment (jleveque) [inline src/system-health/scripts/healthd:106]: `thermal_control`? Copy/paste error?

**Review by jleveque — CHANGES_REQUESTED** [review]

File: src/system-health/scripts/healthd
Comment (Junchao-Mellanox) [reply src/system-health/scripts/healthd]: Fixed.

File: src/system-health/scripts/healthd
Comment (jleveque) [inline src/system-health/scripts/healthd:64]: Signal handler should be inherited from DaemonBase. Unless you want to override it, this should not be necessary.

File: src/system-health/scripts/healthd
Comment (jleveque) [inline src/system-health/scripts/healthd:18]: This logger is unnecessary. DaemonBase class now inherits from Logger, so HealthDaemon class should have built-in logging functionality. See this PR for examples: https://github.com/Azure/sonic-buildimage/pull/5370

**Review by jleveque — CHANGES_REQUESTED** [review]

File: src/system-health/scripts/healthd
Comment (Junchao-Mellanox) [reply src/system-health/scripts/healthd:64]: We override it by set a stop event here. And after the event was set, the infinite loop will exit and do some clear operation.

File: src/system-health/scripts/healthd
Comment (Junchao-Mellanox) [reply src/system-health/scripts/healthd]: Fixed

File: src/system-health/scripts/healthd
Comment (jleveque) [inline src/system-health/scripts/healthd:69]: Note that the default log priority of the Logger class is currently NOTICE. If you want these INFO level logs to get logged, you will need to set the minimum log priority to INFO using `set_min_log_priority_info()` like is done here: https://github.com/Azure/sonic-buildimage/pull/5370/files#diff-fd5407b6f3a90e41c17bb90532f90bd1R218

**Review by jleveque — DISMISSED** [review]
Looks good to me. Please wait for other reviewers.

**Review by keboliu — APPROVED** [review]

**Comment by keboliu** [comment]
@sujinmkang, does this look ok for you? if yes, would please approve?

**Review by sujinmkang — APPROVED** [review]

**Comment by keboliu** [comment]
retest this please

**Comment by Junchao-Mellanox** [comment]
retest vs please

**Review by sujinmkang — DISMISSED** [review]

**Review by liat-grozovik — DISMISSED** [review]
and it comes with tests :-)

**Comment by liat-grozovik** [comment]
@Junchao-Mellanox could you please resolve conflicts and once all tests pass we can merge

**Comment by Junchao-Mellanox** [comment]
@liat-grozovik resolved.

**Review by keboliu — APPROVED** [review]

**Review by liat-grozovik — APPROVED** [review]

**Pull request merged**
